server/client.go

```diff
@@ -310,10 +310,19 @@ func (c *client) processErr(errStr string) {
 func (c *client) processConnect(arg []byte) error {
 	c.traceInOp("CONNECT", arg)
 
-	// This will be resolved regardless before we exit this func,
-	// so we can just clear it here.
 	c.mu.Lock()
-	c.clearAuthTimer()
+	// If we can't stop the timer because the callback is in progress...
+	if !c.clearAuthTimer() {
+		// wait for it to finish and handle sending the failure back to
+		// the client.
+		for c.nc != nil {
```

Comment:
I would rather wait on some kind of waitgroup here, but I think that we start to be conservative with what we add in the client structure.
This situation (unable to stop the timer because in progress) is pretty rare: I have not been able to reproduce with a test case (without changing the code to sleep in proper places). So that simple approach without added object may be ok.